server/src/main/java/io/druid/server/security/Authenticator.java

```diff
@@ -0,0 +1,121 @@
+   * this Filter should skip any authentication checks and proceed to the next Filter.
+   * <p>
+   * If the authentication fails, the Filter should not send an error response. The error response will be sent
+   * after all Filters in the authentication filter chain have been checked.
+   * <p>
+   * If an anonymous request is received, the Filter should continue on to the next Filter, the challenge response
+   * will be sent after the filter chain is exhausted.
+   *
+   * @return Filter that authenticates HTTP requests
+   */
+  @Override
+  public Filter getFilter();
+
+  /**
+   * Return a WWW-Authenticate challenge scheme string appropriate for this Authenticator's authentication mechanism.
+   * <p>
+   * For example, a Basic HTTP implementation should return "Basic", while a Kerberos implementation would return
+   * "Negotiate".
+   *
+   * @return Authentication scheme
+   */
+  public String getAuthChallengeHeader();
+
+  /**
+   * Given a JDBC connection context, authenticate the identity represented by the information in the context.
+   * This is used to secure JDBC access for Druid SQL.
+   * <p>
+   * For example, a Basic HTTP auth implementation could read the "user" and "password" fields from the JDBC context.
+   * <p>
+   * The expected contents of the context are left to the implementation.
+   *
+   * @param context JDBC connection context
+   *
+   * @return true if the identity represented by the context is successfully authenticated
+   */
+  public AuthenticationResult authenticateJDBCContext(Map<String, Object> context);
+
+  /**
+   * Return a client that sends requests with the format/information necessary to authenticate successfully
+   * against this Authenticator's authentication scheme using the identity of the internal system user.
+   * <p>
+   * This HTTP client is used for internal communications between Druid nodes, such as when a broker communicates
+   * with a historical node during query processing.
+   *
+   * @param baseClient Base HTTP client for internal Druid communications
+   *
+   * @return metamx HttpClient that sends requests with the credentials of the internal system user
+   */
+  public HttpClient createEscalatedClient(HttpClient baseClient);
+
+  /**
+   * Return a client that sends requests with the format/information necessary to authenticate successfully
+   * against this Authenticator's authentication scheme using the identity of the internal system user.
+   * <p>
+   * This HTTP client is used by the Druid Router node.
+   *
+   * @param baseClient Base Jetty HttpClient
+   *
+   * @return Jetty HttpClient that sends requests with the credentials of the internal system user
+   */
+  public org.eclipse.jetty.client.HttpClient createEscalatedJettyClient(org.eclipse.jetty.client.HttpClient baseClient);
+
+  /**
+   * @return an AuthenticationResult representing the identity of the internal system user.
+   */
+  public AuthenticationResult createEscalatedAuthenticationResult();
```

Comment:
is this only used by sql module to access code in QueryLifecycle to refresh segment metadata etc? shouldn't that be an "allowed all" situation ?

Reply:
I had it as an "allow all" initially, but changed it to address this comment:

https://github.com/druid-io/druid/pull/4271#discussion_r133595627

Reply:
@gianm @jon-wei i see the point about minimizing extra code paths and having some identity written to logs etc. How about having a singleton AuthenticationResult instance which has fixed literal `__druid_internal_superuser` identity and `allow-all` Authorizer . We can use this AuthenticationResult in situations like this and its transparent to both the users and extension writers.

Reply:
I'm open to requiring a common, fixed literal for the internal superuser, but if it's directed to "allowAll", that would require the "allowAll" authorizer to always be instantiated, and I think it's better to have such requests checked by an authorizer that the user actually configured (e.g., maybe the user has special auditing/logging functionality in their authorizer)

Reply:
I like the idea of a reserved name for the superuser.

I'm not sure if it's better to run it through the user's authorizer or not. It would mean that all authorizers have to start with some block that's like "if this is the superuser, then allow it". It does give them more control though. I don't have a strong feeling one way or the other.

Reply:
The user authorizer would have to be able to recognize the internal superuser anyway for requests generated using the HTTP client returned by Authenticator.getEscalatedClient().

In my mind, the internal superuser isn't really a special identity that authorizers would need a special code block for, it should be handled like any other identity, just with permissions appropriate for an internal superuser in whatever permissions system an authorizer decides to use.

Reply:
what is wrong with "allowAll" Authorizer to always exist ? so for superuser that authorizer would be used and other Authorizers don't need to have anything special.
and if we want to make superuser Authorizer configurable then that should be a separate runtime property like "internalAuthorizerType=blah" (default = "allowAll") rather than having additional method on Authenticator interface.

does that work ?

Reply:
Backing up a bit...

To secure internally generated requests sent from one Druid node to another, we need an "escalated identity" for such internode communications, this is the identity used by `Authenticator.getEscalatedClient()`.

The `druid.hadoop.security.kerberos.principal` in the existing Kerberos extension is the same concept:
 http://druid.io/docs/latest/development/extensions-core/druid-kerberos.html

We can't impose requirements that the internode escalated identity have a specific fixed string (some authentication schemes like Kerberos have specific format requirements for the principal string: https://web.mit.edu/kerberos/krb5-1.5/krb5-1.5.4/doc/krb5-user/What-is-a-Kerberos-Principal_003f.html).

Given that such an identity must already exist and be recognized by Authorizers, my stance is to reuse this identity for this situation where we need an escalated AuthenticationResult, and feed it into the same Authorizer that would check escalated internode communications as well, instead of creating another special "internal superuser" here with special handling regarding what Authorizer it goes to.

The implementation of `getEscalatedAuthenticationResult` should be very simple, the Authenticator can just put in the same identity string that would be used in requests generated by the  `getEscalatedClient()`client, and attach the same Authorizer name it would attach if the Authenticator were to handle an escalated internode request from another druid node.

Reply:
thanks for the explanation. I can see why it makes sense, let us keep it.